Normative: Fix ZonedDateTime::round incorrect results with smallestUnit:day

arshaw opened the pull request:
This is my proposed fix to #2790

Adjusted tests:
https://github.com/tc39/test262/compare/main...fullcalendar:test262:temporal-2790-round-zdt-day

Could someone please review?

ptomato commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -5177,7 +5164,7 @@ export function RoundTime(
       quantity = quantity.multiply(1000).plus(nanosecond);
   }
-  const nsPerUnit = unit === 'day' ? dayLengthNs : nsPerTimeUnit[unit];
+  const nsPerUnit = unit === 'day' ? 86400e9 : nsPerTimeUnit[unit];
```
Could probably just fold this into a new entry in `nsPerTimeUnit` if it isn't variable now.

ptomato commented on polyfill/lib/zoneddatetime.mjs:
```diff
@@ -393,8 +393,13 @@ export class ZonedDateTime {
     ]);
-    const offsetNs = ES.GetOffsetNanosecondsFor(timeZoneRec, GetSlot(this, INSTANT));
-    const dt = ES.GetPlainDateTimeFor(timeZoneRec, GetSlot(this, INSTANT), GetSlot(this, CALENDAR), offsetNs);
+
+    const calendar = GetSlot(this, CALENDAR);
+    const instant = GetSlot(this, INSTANT);
+    const thisNs = GetSlot(instant, EPOCHNANOSECONDS);
+
+    const offsetNs = ES.GetOffsetNanosecondsFor(timeZoneRec, instant);
+    const dt = ES.GetPlainDateTimeFor(timeZoneRec, instant, calendar, offsetNs);
```
Mind splitting this out to an editorial PR? As far as I can tell, DRY-ing the `GetSlot` calls isn't observable. Delegates reviewing for TC39 are going to want to see the minimal set of normative changes.

ptomato reviewed (approved):
Thanks for doing this, and the test262 tests as well! I'm pretty sure I understand it and it seems correct to me. Good to know that there were actually a few wrong tests in test262, not just adjustments to the observable user calls.

To present this to TC39 we'll need a change to the spec text. Are you interested in taking on that task as well?

arshaw commented:
@ptomato, sorry for the delay. I was on vaca for 1 week and upon returning have lots to catch up on. Will try my best to address all feedback by EOW.

ptomato commented:
> @ptomato, sorry for the delay. I was on vaca for 1 week and upon returning have lots to catch up on. Will try my best to address all feedback by EOW.

No problem. I think we are still in plenty of time for the TC39 agenda review deadline, 2024-03-29. If you would like help with the spec text, I'd be happy to contribute that part if it will let you concentrate on catching up with other things.

arshaw commented:
Hi @ptomato, I addressed all feed back and also added some tests that will make all new code have coverage.

Writing the spec text was driving me nuts as I'm unfamiliar with the notation and conventions. I'd greatly appreciate it if you could do it.

I'm happy to do any further refinements to make this compatible with what TC39 expects.

ptomato commented:
@arshaw I've got a draft of the spec text, would you mind checking the "Allow edits from maintainers" checkbox? Then I can push it to this branch in a new commit.

Could you also submit the relevant test262 tests in fullcalendar/test262 as a PR to tc39/test262? I would offer to do it and save you some time, but then the UI won't let me approve my own PR.

ptomato commented:
That's odd - it specifically shouldn't be necessary for me to be a collaborator on the source repo and that's never happened to me before! But once I accepted the invitation, I was able to add the commit.

Sorry about that!

arshaw commented:
@ptomato, I built and reviewed your spec changes, and everything looks as expected. Thank you so much.

gibson042 reviewed (approved):
LGTM!

ptomato commented:
This change reached consensus at the TC39 meeting of 2024-04-08.

ptomato commented:
Tests failed due to the formatting snapshot of `chinese` and `dangi` calendars having changed in whatever Node point release is being installed on CI. That's a test in test262 staging which should probably not be there. I'll look into that tomorrow.

ptomato commented:
test262 fix is in https://github.com/tc39/test262/pull/4048. If it gets merged within a few days, I will update this PR to pull it in. If not, I'll add that test to the expected failures file and merge this.
